Add support for removing unsupported huge page sizes

odinuge opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**What type of PR is this?**
> Uncomment only one ` /kind <>` line, hit enter to put that in a new line, and remove leading whitespaces from that line:
>

/kind bug


**What this PR does / why we need it**:
When kubelet is restarted, it will now remove the resources for huge
page sizes no longer supported. This is required when:
- node disables huge pages
- changing the default huge page size in older versions of linux
(because it will then only support the newly set default).
- Software updates that change what sizes are supported (eg. by changing
boot parameters).

In the same way as other resource types, we should also do a cleanup when given resources are no longer supported. The scheduler will handle this without a problem. There should be no backwards compatibility issues, since this would previously result in a crash in kubelet.

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #80452

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
-->
```release-note
Fix issues when supported huge page sizes changes
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.
Please use the following format for linking documentation or pass the
section below:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
- [KEP]: https://github.com/kubernetes/enhancements/tree/master/keps/sig-node
- [Usage]: https://kubernetes.io/docs/tasks/manage-hugepages/scheduling-hugepages/
```

odinuge committed:
Add support for removing unsupported huge page sizes

When kubelet is restarted, it will now remove the resources for huge
page sizes no longer supported. This is required when:
- node disables huge pages
- changing the default huge page size in older versions of linux
(because it will then only support the newly set default).
- Software updates that change what sizes are supported (eg. by changing
boot parameters).

odinuge commented:
/sig node
/assign @derekwaynecarr 
/cc @mattjmcnaughton

mattjmcnaughton reviewed (approved):
/lgtm
/assign @random-liu

odinuge commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -129,6 +130,25 @@ func (kl *Kubelet) tryRegisterWithAPIServer(node *v1.Node) bool {
+// reconcileHugePageSizes will update huge page capacity for each page size and remove huge page sizes no longer supported
+func (kl *Kubelet) reconcileHugePageSizes(initialNode, node *v1.Node) bool {
+	requiresUpdate := false
+	for k := range node.Status.Capacity {
```
Heh, kinda agree! Will fix! 😄

odinuge commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -129,6 +130,25 @@ func (kl *Kubelet) tryRegisterWithAPIServer(node *v1.Node) bool {
+
+		// If huge page size no longer is supported, we remove it from the new node
+		if _, initialNodeHasResource := initialNode.Status.Capacity[k]; !initialNodeHasResource {
+			delete(node.Status.Capacity, k)
```
"yes". It happens automatically now, so I thought this was enough.

Didn't think about writing a test for it, but that was smart! Will do!

odinuge replied:
Added a new test for it now. :smile:

odinuge reviewed (commented):
Thanks for the review @mattjmcnaughton! Will fix it in a few hours

oomichi commented on test/e2e_node/hugepages_test.go:
```diff
@@ -28,10 +28,12 @@ import (
 	"k8s.io/kubernetes/test/e2e/framework"
 	e2elog "k8s.io/kubernetes/test/e2e/framework/log"
 	e2epod "k8s.io/kubernetes/test/e2e/framework/pod"
+
```
nit: This new blank line would be unnecessary because importing the same "k8s.io/kubernetes"

oomichi commented on test/e2e_node/hugepages_test.go:
```diff
@@ -173,6 +175,32 @@ func runHugePagesTests(f *framework.Framework) {
+		ginkgo.By("mimicking support for 9Mi of 3Mi huge page memory by patching the node status")
+		patch := []byte(`[{"op": "add", "path": "/status/capacity/hugepages-3Mi", "value": "9Mi"}, {"op": "add", "path": "/status/allocatable/hugepages-3Mi", "value": "9Mi"}]`)
+		result := f.ClientSet.CoreV1().RESTClient().Patch(types.JSONPatchType).Resource("nodes").Name(framework.TestContext.NodeName).SubResource("status").Body(patch).Do()
+
```
nit: unnecessary blank line between result and its check.

oomichi reviewed (commented):
@odinuge Most part seems good for me, will you update @mattjmcnaughton's points?

oomichi commented:
/cc @oomichi

odinuge committed:
Add support for stopping kubelet in node-e2e

This makes it possible to stop the kubelet, do some work, and then
start it again.

odinuge commented:
/retest

odinuge commented:
/retest

odinuge commented:
/retest

odinuge commented:
I have fixes the comments now, so this is ready for review. :smile: 

cc @oomichi @mattjmcnaughton @derekwaynecarr

mattjmcnaughton reviewed (commented):
Cool!

One high level thought - we have great e2e test coverage here imo... is there benefit to also adding unit test coverage?

Other than that, lgtm but will let someone with slightly more huge pages experience approve :)

odinuge commented:
> Cool!
>
> One high level thought - we have great e2e test coverage here imo... is there benefit to also adding unit test coverage?
>
> Other than that, lgtm but will let someone with slightly more huge pages experience approve :)

Thanks for the suggestion @mattjmcnaughton. Added a unit test as well now.

cc @derekw

odinuge commented:
Ref other huge page issues, I guess you may be interested in this.

/cc @ConnorDoyle 

(also, will rebase when I get some feedback from sig-node :smile: )

odinuge committed:
Add verbose message when more than one kubelet is running

odinuge commented:
ping @kubernetes/sig-node-pr-reviews

odinuge commented:
/retest

bart0sh commented on test/e2e_node/util.go:
```diff
@@ -418,18 +418,34 @@ func getCRIClient() (internalapi.RuntimeService, internalapi.ImageManagerService
+// stopKubelet will kill the running kubelet, and returns a func that will restart the process again
+func stopKubelet() func() {
+	kubeletServiceName := findRunningKubletServiceName()
+	stdout, err := exec.Command("sudo", "systemctl", "kill", kubeletServiceName).CombinedOutput()
```
I'd suggest to use ```systemctl stop``` here. Unlike ```systemctl kill``` it will use official configured way to stop the service.

odinuge replied:
When using `systemctl stop` the systemd service will be deleted, making it impossible to start it again.

> If killing kubelet is mandatory for your tests then it would be better to call this function killKubelet.

Not sure what you mean with `killKubelet`.

bart0sh replied:
> When using systemctl stop the systemd service will be deleted, making it impossible to start it again.

Can you elaborate on this? Not sure I follow.

> Not sure what you mean with killKubelet.

I mean that you called function stopKubelet, but it doesn't stop Kubelet, it kills it. I proposed to call it KillKubelet.

odinuge replied:
> Can you elaborate on this? Not sure I follow.

```bash
# Using stop
> systemd-run --remain-after-exit sleep 100
Running as unit: run-u14192.service
> systemctl stop run-u14192.service
> systemctl start run-u14192.service
Failed to start run-u14192.service: Unit run-u14192.service not found.

# Using kill                                                                                                                                                                                             
> systemd-run --remain-after-exit sleep 100
Running as unit: run-u14224.service 
> systemctl kill run-u14224.service
> systemctl start run-u14224.service
# service is restarted
```

> I mean that you called function stopKubelet, but it doesn't stop Kubelet, it kills it. I proposed to call it KillKubelet.

Aha, I agree that renaming it would make sense. I thought you meant to call a given function named `killKubelet` instead.

derekwaynecarr replied:
we can rename the function in a follow-on, but i don't see a major reason to block on the core issue addressed in the pr.

bart0sh commented on test/e2e_node/util.go:
```diff
@@ -418,18 +418,34 @@ func getCRIClient() (internalapi.RuntimeService, internalapi.ImageManagerService
 }
 
+// stopKubelet will kill the running kubelet, and returns a func that will restart the process again
+func stopKubelet() func() {
```
where this function is used and how is it related to this PR?

odinuge commented on test/e2e_node/hugepages_test.go:
```diff
@@ -167,12 +168,57 @@ func runHugePagesTests(f *framework.Framework) {
+
+	ginkgo.It("should add resources for new huge page sizes on kubelet restart", func() {
+		ginkgo.By("Stopping kubelet")
+		startKubelet := stopKubelet()
```
> @bart0sh where this function is used and how is it related to this PR?

It is used here

bart0sh replied:
Can you explain why stopKubelet function returns another function that starts Kubelet? It doesn't look very clear and logical to me to be honest.

odinuge replied:
If you look at the func it make more sense. We have to store the name of the service somewhere, and imo. it is cleaner to use a closure like this instead of returning the name, or even storing it in a var.

If you have a better idea ill be happy to change the behavior! :)

bart0sh replied:
I'd say storing it in var is much more clear, readable and allows to check if the action succeeded:
```
kubeletServiceName := findRunningKubletServiceName()
err := killKubelet(kubeletServiceName)
...
err := startKubelet(kubeletServiceName)
```

odinuge replied:
I guess we can do that, but then we will have to check with the folks from `sig-testing`. 

Doing that will move the logic for handling the different background services into the test, making it hard to refactor the code later. In case we start supporting other ways to run kubelet in the tests than `systemd`, we will have to change the tests too, not just the utilities.

I do think I prefer to separate the two concerns, but ill defer the decision to `sig-testing`.

cc @oomichi

bart0sh replied:
I think either me or you misunderstood something. I'm not proposing to support other ways to run kubelet. My proposal is to get rid of closure.
Here is how would killKubelet functions would look like:

```
func killKubelet(kubeletServiceName string) error {
	stdout, err := exec.Command("sudo", "systemctl", "kill", kubeletServiceName).CombinedOutput()
         framework.ExpectNoError(err, "Failed to kill kubelet with systemctl: %v, %v", err, stdout)
         return err
}

```

odinuge replied:
> I'm not proposing to support other ways to run kubelet. My proposal is to get rid of closure.

Adding this code to the test will imo. move the logic of handlig the liveness of kubelet into the test, instead of a central & reusable place.
```go
kubeletServiceName := findRunningKubletServiceName()
err := killKubelet(kubeletServiceName)
...
err := startKubelet(kubeletServiceName)
````

I do agree that `no closure  > closure`, but I am not that happy with moving the systemd-logic into the test. But again, ill defer to sig-testing.

Long term the best solution would be to expose these `cmd`s into the `e2e_node/utils`, but that is outside the scope of this PR.
https://github.com/kubernetes/kubernetes/blob/master/test/e2e_node/services/kubelet.go#L183-L210

bart0sh replied:
I'm still not sure I fully understand the concern, i.e. why this code doesn't expose handling liveness of kubelet into the test:
```
startKubelet := killKubelet()
...
startKubelet()
```

but this one does:
```
kubeletServiceName := findRunningKubletServiceName()
err := killKubelet(kubeletServiceName)
...
err := startKubelet(kubeletServiceName)
```

However, I still think the latter is much more readable and understandable than the former.

odinuge commented:
/test pull-kubernetes-e2e-gce

odinuge commented:
/retest

odinuge commented:
/retest

odinuge commented:
/retest

tedyu commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -127,6 +129,53 @@ func (kl *Kubelet) tryRegisterWithAPIServer(node *v1.Node) bool {
 	return true
 }
 
+// reconcileHugePageSizes will update huge page capacity for each page size and remove huge page sizes no longer supported
```
reconcileHugePageSizes  -> reconcileHugePageResource

cynepco3hahue commented:
@odinuge Can you please rebase it, and does it possible to get it into 1.18?

odinuge commented:
/retest  pull-kubernetes-integration

odinuge commented:
/test pull-kubernetes-integration

derekwaynecarr commented on pkg/kubelet/kubelet_node_status.go:
```diff
@@ -117,6 +118,7 @@ func (kl *Kubelet) tryRegisterWithAPIServer(node *v1.Node) bool {
 	requiresUpdate := kl.reconcileCMADAnnotationWithExistingNode(node, existingNode)
 	requiresUpdate = kl.updateDefaultLabels(node, existingNode) || requiresUpdate
 	requiresUpdate = kl.reconcileExtendedResource(node, existingNode) || requiresUpdate
+	requiresUpdate = kl.reconcileHugePageResource(node, existingNode) || requiresUpdate
```
I need to catch-up on latest state of this code, but we traditionally do not support hot-swap of other resources like cpu or memory for a same named node without a restart.  i thought i had seen a similar pr with an update in this space last year, and so I need to refresh my own memory if we handled hugepages separately.

derekwaynecarr replied:
well, devices would get updated as a counted resource.

derekwaynecarr replied:
@odinuge ok, this makes sense.

derekwaynecarr commented on test/e2e_node/hugepages_test.go:
```diff
@@ -178,12 +179,57 @@ func runHugePagesTests(f *framework.Framework) {
 var _ = SIGDescribe("HugePages [Serial] [Feature:HugePages][NodeFeature:HugePages]", func() {
 	f := framework.NewDefaultFramework("hugepages-test")
 
+	ginkgo.It("should remove resources for huge page sizes no longer supported", func() {
```
this is a rare test that we can exercise with hugepages in shared testing infrastructure as it does not depend on native support in the test environment.

derekwaynecarr reviewed (commented):
I need to double check latest state of code when I have my editor present, but this looks mostly good to me.

cynepco3hahue commented:
Does it anything that prevents it to be merged before 1.19 code freeze?

derekwaynecarr reviewed (approved):
/approve
/lgtm

i am open to picking this back to prior releases if would like as it addresses a bug where users may allocate huge pages via privileged daemonset and then signal restart kubelet.

odinuge commented:
/retest
